crates/sound/src/vhu_sound.rs

```diff
@@ -192,42 +218,51 @@ impl VhostUserSoundThread {
-                        pcm.direction = VIRTIO_SND_D_OUTPUT;
-                        pcm.channels_min = 1;
-                        pcm.channels_max = 6;
-                        pcm.padding = [0; 5];
-                    }
-                    if start_id + count > pcm_info.len() {
+
+                    if start_id + count > stream_info.len() {
                         error!(
                             "start_id({}) + count({}) must be smaller than the number of streams ({})",
                             start_id,
                             count,
-                            pcm_info.len()
+                            stream_info.len()
                         );
                         desc_chain
                             .memory()
                             .write_obj(VIRTIO_SND_S_BAD_MSG, desc_response.addr())
                             .map_err(|_| Error::DescriptorWriteFailed)?;
-                    }
-                    desc_chain
-                        .memory()
-                        .write_obj(response, desc_response.addr())
-                        .map_err(|_| Error::DescriptorWriteFailed)?;
-                    //let mut len = desc_response.len() as u32;
+                    } else {
+                        desc_chain
+                            .memory()
+                            .write_obj(response, desc_response.addr())
+                            .map_err(|_| Error::DescriptorWriteFailed)?;
 
-                    for i in start_id..(start_id + count) {
+                        let mut buf = vec![];
+
+                        for i in start_id..start_id+count {
+                            let pcm_info = VirtioSoundPcmInfo {
+                                hdr : VirtioSoundInfo {
+                                    hda_fn_nid : Le32::from(i as u32)
+                                },
+                                features : stream_info[i].features,
+                                formats : stream_info[i].formats,
+                                rates : stream_info[i].rates,
+                                direction : stream_info[i].direction,
+                                channels_min : stream_info[i].channels_min,
+                                channels_max : stream_info[i].channels_max,
+                                padding : [0; 5]
+                            };
+                            buf.extend_from_slice(pcm_info.as_slice());
+                        };
+
+                        // TODO: to support the case when the number of items
+                        // do not fit in a single descriptor
                         desc_chain
                             .memory()
-                            .write_slice(pcm_info[i].as_slice(), desc_pcm.addr())
+                            .write_slice(&buf, desc_pcm.addr())
                             .map_err(|_| Error::DescriptorWriteFailed)?;
+
+                        len += desc_pcm.len();
```

Comment:
Not for this patch, but I would rename `len` in `used_len`, be sure to increment it every time we write a descriptor, then call `.add_used` once at the end, so we will cover both error paths and normal paths.

WDYT?

Reply:
I agree. I will do it in a following PR.